File: app/src/main/java/com/callmangement/tracking_service/GetLocation.java
package com.callmangement.tracking_service;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.location.Address;
import android.location.Geocoder;
import android.location.Location;
import android.location.LocationManager;
import android.os.Handler;

import com.callmangement.database.DbController;
import com.callmangement.utils.DateTimeUtils;
import com.callmangement.utils.MyDialog;
import com.callmangement.utils.PrefManager;
import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.location.LocationSettingsRequest;
import com.google.android.gms.location.LocationSettingsResponse;
import com.google.android.gms.tasks.Task;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.Objects;

import static android.content.Context.LOCATION_SERVICE;
import static com.callmangement.utils.Constants.currentLat;
import static com.callmangement.utils.Constants.currentLong;
import static com.callmangement.utils.Constants.currentTime;
import static com.callmangement.utils.MyDialog.alertD;

public class GetLocation {
    private LocationRequest mLocationRequest;
    private final Context mContext;
    private final PrefManager prefManager;

    @SuppressLint("MissingPermission")
    public GetLocation(Context context) {
        mContext = context;
        prefManager = new PrefManager(context);
    }

    public void build() {
        createLocationRequest();
        checkLocationSettings();
    }

    private void createLocationRequest() {
        mLocationRequest = LocationRequest.create();
        mLocationRequest.setInterval(60000);
        mLocationRequest.setFastestInterval(60000);
        mLocationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
    }

    private void checkLocationSettings() {
        LocationSettingsRequest.Builder builder = new LocationSettingsRequest.Builder().addLocationRequest(mLocationRequest).setAlwaysShow(false);
        initializeTimerTask(builder);
    }

    public void initializeTimerTask(final LocationSettingsRequest.Builder builder) {
        Task<LocationSettingsResponse> result = LocationServices.getSettingsClient(Objects.requireNonNull(mContext)).checkLocationSettings(builder.build());
        result.addOnCompleteListener(task -> {
            try {
                LocationManager locationManager = (LocationManager) mContext.getSystemService(LOCATION_SERVICE);
                if (locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)) {
                    getLocation();
                } else if (locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER)) {
                    getLocation();
                } else {
                    Runnable runnable = () -> {
                        if (alertD == null || !alertD.isShowing()) {
                            Intent intent = new Intent(mContext, MyDialog.class);
                            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                            mContext.startActivity(intent);
                        }
                    };
                    Handler handler = new Handler();
                    handler.postDelayed(runnable, 1500);
                }

            } catch (Exception exception) {
                exception.printStackTrace();
            }
        });
    }

    @SuppressLint("MissingPermission")
    private void getLocation() {
        FusedLocationProviderClient mFusedLocationClient = LocationServices.getFusedLocationProviderClient(Objects.requireNonNull(mContext));
        Task<Location> location = mFusedLocationClient.getLastLocation();
        location.addOnSuccessListener(location1 -> {
            @SuppressLint("SimpleDateFormat") SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss aa");
            double lat1, lon1, lat2, lon2;
            Date date1;
            Date date2;
            long diffSeconds = 0;
            double speed = 0;
            if (location1 != null) {
                currentLat = prefManager.getUserCurrentlat();
                currentLong = prefManager.getUserCurrentlong();
                currentTime = prefManager.getUserCurrenttime();
                if ((currentLat == null || currentLat.equalsIgnoreCase("")) && ((currentLong == null || currentLong.equalsIgnoreCase("")))) {
                    currentLat = String.valueOf(location1.getLatitude());
                    currentLong = String.valueOf(location1.getLongitude());
                    Date c = Calendar.getInstance().getTime();
                    currentTime = simpleDateFormat.format(c);

                }
                else {
                    lat1 = Double.parseDouble(Objects.requireNonNull(currentLat));
                    lon1 = Double.parseDouble(currentLong);
                    lat2 = location1.getLatitude();
                    lon2 = location1.getLongitude();
                    //current time
                    Date c = Calendar.getInstance().getTime();
                    String currentTimeNew = simpleDateFormat.format(c);
                    try {
                        date1 = simpleDateFormat.parse(currentTime);
                        date2 = simpleDateFormat.parse(currentTimeNew);
                        //find difference
                        long diff = Objects.requireNonNull(date2).getTime() - Objects.requireNonNull(date1).getTime();
                        diffSeconds = diff / 1000;
                        long diffMinutes = diff / (60 * 1000) % 60;
                        long diffHours = diff / (60 * 60 * 1000) % 24;
                        long diffDays = diff / (24 * 60 * 60 * 1000);
                        long difonlysec = diff / 1000;
                    } catch (ParseException e) {
                        e.printStackTrace();
                    }
                    //Calculate distance
                    double earthRadius = 6371000; //meters
                    double dLat = Math.toRadians(lat2 - lat1);
                    double dLng = Math.toRadians(lon2 - lon1);
                    double a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
                            Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2)) *
                                    Math.sin(dLng / 2) * Math.sin(dLng / 2);
                    double cc = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
                    float dist = (float) (earthRadius * cc);
                    float distMeter = dist * 1000;
                    //calculate speed
                    speed = distMeter / diffSeconds;
                    currentLat = String.valueOf(lat2);
                    currentLong = String.valueOf(lon2);
                    currentTime = currentTimeNew;
                }
                prefManager.setUserCurrentlat(currentLat);
                prefManager.setUserCurrentlong(currentLong);
                prefManager.setUserCurrenttime(currentTime);
                //calculate speed
//                    if (speed < 150000) {
//                        updateCityAndPincode(location.getLatitude(), location.getLongitude());
//                    }
                updateCityAndPincode(location1.getLatitude(), location1.getLongitude());
            }
        });
    }
    private void updateCityAndPincode(double latitude, double longitude) {
        try {
            Geocoder gcd = new Geocoder(mContext, Locale.getDefault());
            List<Address> addresses = gcd.getFromLocation(latitude, longitude, 1);
            if (addresses.size() > 0) {
                String add1 = "" + addresses.get(0).getAddressLine(0);
                if (add1.length() > 0) {
                    saveLocation(add1, latitude, longitude);
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void saveLocation(String address, double latitude, double longitude) {
        //Log.e("location","latitude - "+latitude +", longitude - "+longitude+", address - "+ address);
        new DbController(mContext).insertLocation(prefManager.getUseR_Id(),prefManager.getUseR_DistrictId(),String.valueOf(latitude), String.valueOf(longitude), address, DateTimeUtils.getCurrentDataTime());
    }

}

File: app/src/main/java/com/callmangement/tracking_service/GetLocation.kt
package com.callmangement.tracking_service

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.location.Geocoder
import android.location.Location
import android.location.LocationManager
import android.os.Handler
import com.callmangement.database.DbController
import com.callmangement.utils.Constants
import com.callmangement.utils.DateTimeUtils
import com.callmangement.utils.MyDialog
import com.callmangement.utils.PrefManager
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationServices
import com.google.android.gms.location.LocationSettingsRequest
import com.google.android.gms.location.LocationSettingsResponse
import com.google.android.gms.tasks.Task
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.Objects
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

class GetLocation @SuppressLint("MissingPermission") constructor(private val mContext: Context) {
    private var mLocationRequest: LocationRequest? = null
    private val prefManager = PrefManager(mContext)

    fun build() {
        createLocationRequest()
        checkLocationSettings()
    }

    private fun createLocationRequest() {
        mLocationRequest = LocationRequest.create()
        mLocationRequest!!.setInterval(60000)
        mLocationRequest!!.setFastestInterval(60000)
        mLocationRequest!!.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY)
    }

    private fun checkLocationSettings() {
        val builder = LocationSettingsRequest.Builder().addLocationRequest(
            mLocationRequest!!
        ).setAlwaysShow(false)
        initializeTimerTask(builder)
    }

    fun initializeTimerTask(builder: LocationSettingsRequest.Builder) {
        val result = LocationServices.getSettingsClient(Objects.requireNonNull(mContext))
            .checkLocationSettings(builder.build())
        result.addOnCompleteListener { task: Task<LocationSettingsResponse?>? ->
            try {
                val locationManager =
                    mContext.getSystemService(Context.LOCATION_SERVICE) as LocationManager
                if (locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)) {
                    location
                } else if (locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER)) {
                    location
                } else {
                    val runnable = Runnable {
                        if (MyDialog.alertD == null || !MyDialog.alertD!!.isShowing) {
                            val intent = Intent(mContext, MyDialog::class.java)
                            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                            mContext.startActivity(intent)
                        }
                    }
                    val handler = Handler()
                    handler.postDelayed(runnable, 1500)
                }
            } catch (exception: Exception) {
                exception.printStackTrace()
            }
        }
    }

    @get:SuppressLint("MissingPermission")
    private val location: Unit
        get() {
            val mFusedLocationClient =
                LocationServices.getFusedLocationProviderClient(Objects.requireNonNull(mContext))
            val location = mFusedLocationClient.lastLocation
            location.addOnSuccessListener { location1: Location? ->
                @SuppressLint("SimpleDateFormat") val simpleDateFormat =
                    SimpleDateFormat("yyyy-MM-dd HH:mm:ss aa")
                val lat1: Double
                val lon1: Double
                val lat2: Double
                val lon2: Double
                val date1: Date
                val date2: Date
                var diffSeconds: Long = 0
                var speed = 0.0
                if (location1 != null) {
                    Constants.currentLat = prefManager.userCurrentlat!!
                    Constants.currentLong = prefManager.userCurrentlong!!
                    Constants.currentTime = prefManager.userCurrenttime!!
                    if ((Constants.currentLat == null || Constants.currentLat.equals(
                            "",
                            ignoreCase = true
                        )) && ((Constants.currentLong == null || Constants.currentLong.equals(
                            "",
                            ignoreCase = true
                        )))
                    ) {
                        Constants.currentLat = location1.latitude.toString()
                        Constants.currentLong = location1.longitude.toString()
                        val c = Calendar.getInstance().time
                        Constants.currentTime = simpleDateFormat.format(c)
                    } else {
                        lat1 = Objects.requireNonNull(Constants.currentLat).toDouble()
                        lon1 = Constants.currentLong.toDouble()
                        lat2 = location1.latitude
                        lon2 = location1.longitude
                        //current time
                        val c = Calendar.getInstance().time
                        val currentTimeNew = simpleDateFormat.format(c)
                        try {
                            date1 = simpleDateFormat.parse(Constants.currentTime)
                            date2 = simpleDateFormat.parse(currentTimeNew)
                            //find difference
                            val diff =
                                Objects.requireNonNull(date2).time - Objects.requireNonNull(date1).time
                            diffSeconds = diff / 1000
                            val diffMinutes = diff / (60 * 1000) % 60
                            val diffHours = diff / (60 * 60 * 1000) % 24
                            val diffDays = diff / (24 * 60 * 60 * 1000)
                            val difonlysec = diff / 1000
                        } catch (e: ParseException) {
                            e.printStackTrace()
                        }
                        //Calculate distance
                        val earthRadius = 6371000.0 //meters
                        val dLat = Math.toRadians(lat2 - lat1)
                        val dLng = Math.toRadians(lon2 - lon1)
                        val a = sin(dLat / 2) * sin(dLat / 2) + cos(Math.toRadians(lat1)) * cos(
                            Math.toRadians(lat2)
                        ) * sin(dLng / 2) * sin(dLng / 2)
                        val cc = 2 * atan2(sqrt(a), sqrt(1 - a))
                        val dist = (earthRadius * cc).toFloat()
                        val distMeter = dist * 1000
                        //calculate speed
                        speed = (distMeter / diffSeconds).toDouble()
                        Constants.currentLat = lat2.toString()
                        Constants.currentLong = lon2.toString()
                        Constants.currentTime = currentTimeNew
                    }
                    prefManager.userCurrentlat = Constants.currentLat
                    prefManager.userCurrentlong = Constants.currentLong
                    prefManager.userCurrenttime = Constants.currentTime
                    //calculate speed
//                    if (speed < 150000) {
//                        updateCityAndPincode(location.getLatitude(), location.getLongitude());
//                    }
                    updateCityAndPincode(location1.latitude, location1.longitude)
                }
            }
        }

    private fun updateCityAndPincode(latitude: Double, longitude: Double) {
        try {
            val gcd = Geocoder(mContext, Locale.getDefault())
            val addresses = gcd.getFromLocation(latitude, longitude, 1)
            if (addresses!!.size > 0) {
                val add1 = "" + addresses[0].getAddressLine(0)
                if (add1.length > 0) {
                    saveLocation(add1, latitude, longitude)
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun saveLocation(address: String, latitude: Double, longitude: Double) {
        //Log.e("location","latitude - "+latitude +", longitude - "+longitude+", address - "+ address);
        DbController(mContext).insertLocation(
            prefManager.uSER_Id,
            prefManager.uSER_DistrictId,
            latitude.toString(),
            longitude.toString(),
            address,
            DateTimeUtils.currentDataTime
        )
    }
}
